docs/design/libraries/ComInterfaceGenerator/VTableStubs.md

```diff
@@ -0,0 +1,402 @@
+
+```csharp
+namespace System.Runtime.InteropServices;
+
+[AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
+public class VirtualMethodIndexAttribute : Attribute
+{
+    public VirtualMethodIndexAttribute(int index)
+    {
+        Index = index;
+    }
+
+    public int Index { get; }
+
+    public bool ImplicitThisParameter { get; set; } = true;
+
+    public CustomTypeMarshallerDirection Direction { get; set; } = CustomTypeMarshallerDirection.Ref;
+
+    /// <summary>
+    /// Gets or sets how to marshal string arguments to the method.
+    /// </summary>
+    /// <remarks>
+    /// If this field is set to a value other than <see cref="StringMarshalling.Custom" />,
+    /// <see cref="StringMarshallingCustomType" /> must not be specified.
+    /// </remarks>
+    public StringMarshalling StringMarshalling { get; set; }
+
+    /// <summary>
+    /// Gets or sets the <see cref="Type"/> used to control how string arguments to the method are marshalled.
+    /// </summary>
+    /// <remarks>
+    /// If this field is specified, <see cref="StringMarshalling" /> must not be specified
+    /// or must be set to <see cref="StringMarshalling.Custom" />.
+    /// </remarks>
+    public Type? StringMarshallingCustomType { get; set; }
+
+    /// <summary>
+    /// Gets or sets whether the callee sets an error (SetLastError on Windows or errno
+    /// on other platforms) before returning from the attributed method.
+    /// </summary>
+    public bool SetLastError { get; set; }
+}
+
+```
+
+A new interface will be defined and used by the source generator to fetch the native `this` pointer and the vtable that the function pointer is stored in. This interface is designed to provide an API that various native platforms, like COM, WinRT, or Swift, could use to provide support for multiple managed interface wrappers from a single native object. In particular, this interface was designed to ensure it is possible support a managed gesture to do an unmanaged "type cast" (i.e., `QueryInterface` in the COM and WinRT worlds).
+
+```csharp
+namespace System.Runtime.InteropServices;
+
+public readonly ref struct VirtualMethodTableInfo
+{
+    public VirtualMethodTableInfo(IntPtr thisPointer, ReadOnlySpan<IntPtr> virtualMethodTable)
+    {
+        ThisPointer = thisPointer;
+        VirtualMethodTable = virtualMethodTable;
+    }
+
+    public IntPtr ThisPointer { get; }
+    public ReadOnlySpan<IntPtr> VirtualMethodTable { get; }
+
+    public void Deconstruct(out IntPtr thisPointer, out ReadOnlySpan<IntPtr> virtualMethodTable) // This method allows tuple-style `var (thisPtr, vtable) = virtualMethodTableInfo;` statements from this type.
+    {
+        thisPointer = ThisPointer;
+        virtualMethodTable = VirtualMethodTable;
+    }
+}
+
+public interface IUnmanagedVirtualMethodTableProvider<T> where T : IEquatable<T>
+{
+    VirtualMethodTableInfo GetVirtualMethodTableInfoForKey(T typeKey);
```

Comment:
Is the JIT going to be able to de-virtualize this function call in practice? I'd like to see what we can do to avoid adding look-up cost on every function call.

Reply:
It should be able to devirtualize this with Guarded Devirtualization. Since the calling function will be invoked through an IDIC implementation, I don't think the JIT will be able to do the non-guarded devirtualization even in NativeAOT, but I could be wrong.